discovery/eureka_test.go

```diff
@@ -0,0 +1,119 @@
+package discovery
+
+import (
+	"gopkg.in/ory-am/dockertest.v3"
+	dc "github.com/fsouza/go-dockerclient"
+	"github.com/jaume-pinyol/fargo"
+	"log"
+	"os"
+	"testing"
+	"strconv"
+	"github.com/op/go-logging"
+)
+
+var eurekaTestPort int = 8080
+var eurekaTestUrl string = "http://127.0.0.1:" + strconv.Itoa(eurekaTestPort) + "/eureka"
+
+func TestMain(m *testing.M) {
+	// uses a sensible default on windows (tcp/http) and linux/osx (socket)
+	pool, err := dockertest.NewPool("")
+	if err != nil {
+		log.Fatalf("Could not connect to docker: %s", err)
+	}
+	// pulls an image, creates a container based on it and runs it
+	resource, err := pool.RunWithOptions(&dockertest.RunOptions{
+		Repository:   "netflixoss/eureka",
+		Tag:          "1.3.1",
+		PortBindings: map[dc.Port][]dc.PortBinding{
+			dc.Port(strconv.Itoa(eurekaTestPort) + "/tcp"): {{HostIP: "", HostPort: strconv.Itoa(eurekaTestPort)}},
+		},
+	})
+	if err != nil {
+		log.Fatalf("Could not start resource: %s", err)
+	}
+
+	// exponential backoff-retry, because the application in the container might not be ready to accept connections yet
+	if err := pool.Retry(func() error {
+		_, err := NewEurekaClient(eurekaTestUrl)
+		return err
+	}); err != nil {
+		log.Fatalf("Could not connect to the docker resource: %s", err)
+	}
+
+	code := m.Run()
+
+	// You can't defer this because os.Exit doesn't care for defer
+	if err := pool.Purge(resource); err != nil {
+		log.Fatalf("Could not purge resource: %s", err)
+	}
+
+	os.Exit(code)
+}
+
+func TestEurekaClientNoEureka(t *testing.T) {
+	_, err := NewEurekaClient("http://localhost:9999/thisshouldntwork")
+	if err != errNoEurekaConnection {
+		t.Fatal("We shouldnt reach eureka if Eureka hostname/port is completely wrong")
+	}
+}
+
+func TestEurekaClientWrongEurekaContext(t *testing.T) {
+	_, err := NewEurekaClient(eurekaTestUrl + "badsuffix")
+	if err != errEurekaUnexpectedHttpResponseCode {
+		t.Fatal("Eureka should be reachable but, when asking a wrong URL, it should return a non 200 response code")
```

Comment:
mmmm, reaching this point I suspect that the m.Run() in TestMain does is to start the Eureka server to run the other tests after that. am I right?

Reply:
Nops. TestMain does run Eureka (and kills it when tests finish), and in the middle, with m.Run , actually runs all your tests there. Its a mechanism allowing you to wrap all your tests and do things at the beginning and end (init only covers the first part, so that's why golang released this in... 1.5?)

Reply:
👍  seen in your comment above